### PR TITLE
Resurrect AddIndex() and CreateIndex()

### PR DESCRIPTION
**Pull request by glycerine**
Indexes are critical for performance.

In this PR I've resurrected the "index management" pull request below from https://github.com/dmajere and updated it with some documentation in the README.  Feel free to squash/cherrypick/whatever if I haven't prepared it just right, as I'm not the best with git.

https://github.com/go-gorp/gorp/issues/97  was the original.

### CODE REVIEW

File: gorp.go [hunk 1]
```diff
@@ -604,6 +642,115 @@ func (c *ColumnMap) SetMaxSize(size int) *ColumnMap {
...
+
+	// If true, " unique" is added to create index statements.
+	// Not used elsewhere
+	Unique bool
+
+	// Index type supported by Dialect
+	// Postgres:  B-tree, Hash, GiST and GIN.
+	// Mysql: Btree, Hash.
+	// Sqlite: nil.
+	IndexType string
```
Comment (GeertJohan) [inline gorp.go]: The difference between `Btree` and `B-tree` is just the naming right?
So could we find a way to have one identifier that works for both Postgres and MySQL, to make migration between the two seamless when using Btree?

File: gorp.go [hunk 2]
```diff
@@ -604,6 +642,115 @@ func (c *ColumnMap) SetMaxSize(size int) *ColumnMap {
...
+			s.WriteString("create")
+			if index.Unique {
+				s.WriteString(" unique")
+			}
+			s.WriteString(" index")
+			s.WriteString(fmt.Sprintf(" %s on %s", index.IndexName,
+				table.TableName))
+			if dname := dialect.Name(); dname == "PostgresDialect" && index.IndexType != "" {
+				s.WriteString(fmt.Sprintf(" %s %s", m.Dialect.CreateIndexSuffix(), index.IndexType))
+			}
```
Comment (GeertJohan) [inline gorp.go]: As @coopernurse mentioned in https://github.com/go-gorp/gorp/pull/97#issuecomment-30667564
The Dialect's name is inspected here, this behavior should be moved to the Dialect implementations. The whole purpose of the Dialect is to handle the different in SQL Syntaxes.

File: gorp.go [hunk 3]
```diff
@@ -604,6 +642,115 @@ func (c *ColumnMap) SetMaxSize(size int) *ColumnMap {
...
+				if x > 0 {
+					s.WriteString(", ")
+				}
+				s.WriteString(m.Dialect.QuoteField(col))
+			}
+			s.WriteString(")")
+
+			if dname := dialect.Name(); dname == "MySQLDialect" && index.IndexType != "" {
+				s.WriteString(fmt.Sprintf(" %s %s", m.Dialect.CreateIndexSuffix(), index.IndexType))
+			}
```
Comment (GeertJohan) [inline gorp.go]: Same as line 706. Move query generation to Dialect.

File: gorp.go [hunk 4]
```diff
@@ -604,6 +642,115 @@ func (c *ColumnMap) SetMaxSize(size int) *ColumnMap {
...
+	var err error
+	dialect := reflect.TypeOf(t.dbmap.Dialect)
+	for _, idx := range t.indexes {
+		if idx.IndexName == name {
+			s := bytes.Buffer{}
+			s.WriteString(fmt.Sprintf("DROP INDEX %s", idx.IndexName))
+
+			if dname := dialect.Name(); dname == "MySQLDialect" {
+				s.WriteString(fmt.Sprintf(" %s %s", t.dbmap.Dialect.DropIndexSuffix(), t.TableName))
+			}
```
Comment (GeertJohan) [inline gorp.go]: Same as line 706. Move query generation to Dialect.

**Comment by qrawl** [comment]
@glycerine What about adding the (DB-dependant) index type to the `CreateIndexSuffix` function signature, freeing `gorp.go` from inspecting the dialect?

```Go
func (d PostgresDialect) CreateIndexSuffix(indexType string) string {
	return fmt.Sprint("using %s", indexType)
}
```

PS: with #246, indexes (as well as PKs) could be parsed from field tags.

**Comment by GeertJohan** [comment]
This PR needs some work, but it's best to merge in at `v2-phase`` (see #270). Merging after phase2 will be a mess..

**Pull request merged**